### PR TITLE
Modify kernel registration & support fp16

### PR DESCRIPTION
**Pull request by kilinchange**
为避免 support_fp16 分支过于庞大，从 support_fp16 分支拆出来一部分 commit，主要涉及 kernel 注册机制的修改和 llama fp16 的支持修改。后续将在这个分支进行其他平台的 kernel 注册机制适配工作，support_fp16 分支则接着适配量化模型。
具体的修改包括：

框架层面：
- [x] 将 DataType 从 Kernel 注册机制中移除 @kilinchange 
- [x] 修改拓扑排序代码，保证每次运行拓扑序一致 @YdrMaster 
- [x] 优化使用 naive allocator 的方式，直接在初始化 stub 时传 use_naive_allocator 参数选择是否使用 naive 模式 @kilinchange 

算子层面：
- [x] Cpu 算子适配新注册机制 @kilinchange 
- [x] llama 涉及到的 Cuda 算子支持 fp16 计算 @xgqdut2016 @bitzyz @OdinaryWord @kilinchange 
- [x] 其余 Cuda & Bang & intelcpu & kunlun 算子适配新注册机制，算子内部对于不支持的数据类型直接报错 @bitzyz @kilinchange 
- [x] 补充算子：cast @OdinaryWord

### CODE REVIEW

**Comment by kilinchange** [comment]
`cd build/Release && make test
make[1]: Entering directory '/home/duanchenjie/workspace/InfiniTensor/build/Release'
Running tests...
Test project /home/duanchenjie/workspace/InfiniTensor/build/Release
      Start  1: test_graph
 1/71 Test  #1: test_graph .......................   Passed    0.05 sec
      Start  2: test_graph_handler
 2/71 Test  #2: test_graph_handler ...............   Passed    0.05 sec
      Start  3: test_graph_replace
 3/71 Test  #3: test_graph_replace ...............   Passed    0.05 sec
      Start  4: test_hash
 4/71 Test  #4: test_hash ........................   Passed    0.04 sec
      Start  5: test_lazy_allocator
 5/71 Test  #5: test_lazy_allocator ..............   Passed    0.05 sec
      Start  6: test_search
 6/71 Test  #6: test_search ......................   Passed    0.42 sec
      Start  7: test_tensor_save
 7/71 Test  #7: test_tensor_save .................   Passed    0.00 sec
      Start  8: test_verify
 8/71 Test  #8: test_verify ......................   Passed    0.05 sec
      Start  9: test_all_gather
 9/71 Test  #9: test_all_gather ..................   Passed    0.04 sec
      Start 10: test_all_reduce
10/71 Test #10: test_all_reduce ..................   Passed    0.04 sec
      Start 11: test_batch_norm
11/71 Test #11: test_batch_norm ..................   Passed    0.05 sec
      Start 12: test_broadcast
12/71 Test #12: test_broadcast ...................   Passed    0.05 sec
      Start 13: test_clip
13/71 Test #13: test_clip ........................   Passed    0.05 sec
      Start 14: test_concat
14/71 Test #14: test_concat ......................   Passed    0.05 sec
      Start 15: test_conv
15/71 Test #15: test_conv ........................   Passed    0.28 sec
      Start 16: test_conv_transposed_2d
16/71 Test #16: test_conv_transposed_2d ..........   Passed    0.04 sec
      Start 17: test_element_wise
17/71 Test #17: test_element_wise ................   Passed    0.04 sec
      Start 18: test_expand
18/71 Test #18: test_expand ......................   Passed    0.05 sec
      Start 19: test_extend
19/71 Test #19: test_extend ......................   Passed    0.04 sec
      Start 20: test_gather
20/71 Test #20: test_gather ......................   Passed    0.04 sec
      Start 21: test_gather_elements
21/71 Test #21: test_gather_elements .............   Passed    0.05 sec
      Start 22: test_matmul
22/71 Test #22: test_matmul ......................   Passed    0.05 sec
      Start 23: test_pad
23/71 Test #23: test_pad .........................   Passed    0.05 sec
      Start 24: test_pooling
24/71 Test #24: test_pooling .....................   Passed    0.05 sec
      Start 25: test_reduce
25/71 Test #25: test_reduce ......................   Passed    0.05 sec
      Start 26: test_reshape
26/71 Test #26: test_reshape .....................   Passed    0.05 sec
      Start 27: test_resize
27/71 Test #27: test_resize ......................   Passed    0.04 sec
      Start 28: test_sendrecv
28/71 Test #28: test_sendrecv ....................   Passed    0.05 sec
      Start 29: test_slice
29/71 Test #29: test_slice .......................   Passed    0.05 sec
      Start 30: test_split
30/71 Test #30: test_split .......................   Passed    0.05 sec
      Start 31: test_transpose
31/71 Test #31: test_transpose ...................   Passed    0.05 sec
      Start 32: test_unary
32/71 Test #32: test_unary .......................   Passed    0.05 sec
      Start 33: test_where
33/71 Test #33: test_where .......................   Passed    0.05 sec
      Start 34: test_nativecpu_concat
34/71 Test #34: test_nativecpu_concat ............   Passed    0.09 sec
      Start 35: test_nativecpu_split
35/71 Test #35: test_nativecpu_split .............   Passed    0.07 sec
      Start 36: test_nativecpu_transpose
36/71 Test #36: test_nativecpu_transpose .........   Passed    0.04 sec
      Start 37: test_cuda_G2BMM
37/71 Test #37: test_cuda_G2BMM ..................   Passed    2.55 sec
      Start 38: test_cuda_GBMM
38/71 Test #38: test_cuda_GBMM ...................   Passed    3.54 sec
      Start 39: test_cuda_all_gather
39/71 Test #39: test_cuda_all_gather .............   Passed    7.31 sec
      Start 40: test_cuda_all_reduce
40/71 Test #40: test_cuda_all_reduce .............   Passed   13.75 sec
      Start 41: test_cuda_attention
41/71 Test #41: test_cuda_attention ..............   Passed    3.24 sec
      Start 42: test_cuda_batch_norm
42/71 Test #42: test_cuda_batch_norm .............   Passed    0.50 sec
      Start 43: test_cuda_broadcast
43/71 Test #43: test_cuda_broadcast ..............   Passed    7.18 sec
      Start 44: test_cuda_clip
44/71 Test #44: test_cuda_clip ...................   Passed    3.37 sec
      Start 45: test_cuda_concat
45/71 Test #45: test_cuda_concat .................   Passed    3.52 sec
      Start 46: test_cuda_conv
46/71 Test #46: test_cuda_conv ...................   Passed    3.59 sec
      Start 47: test_cuda_conv_fp16
47/71 Test #47: test_cuda_conv_fp16 ..............   Passed    3.51 sec
      Start 48: test_cuda_conv_transposed_2d
48/71 Test #48: test_cuda_conv_transposed_2d .....   Passed    0.79 sec
      Start 49: test_cuda_element_wise
49/71 Test #49: test_cuda_element_wise ...........   Passed    3.43 sec
      Start 50: test_cuda_expand
50/71 Test #50: test_cuda_expand .................   Passed    3.39 sec
      Start 51: test_cuda_extend
51/71 Test #51: test_cuda_extend .................   Passed    3.40 sec
      Start 52: test_cuda_gather
52/71 Test #52: test_cuda_gather .................   Passed    0.60 sec
      Start 53: test_cuda_gather_elements
53/71 Test #53: test_cuda_gather_elements ........   Passed    3.32 sec
      Start 54: test_cuda_inception
54/71 Test #54: test_cuda_inception ..............   Passed    3.54 sec
      Start 55: test_cuda_layernorm
55/71 Test #55: test_cuda_layernorm ..............   Passed    3.53 sec
      Start 56: test_cuda_matmul
56/71 Test #56: test_cuda_matmul .................   Passed    3.61 sec
      Start 57: test_cuda_pad
57/71 Test #57: test_cuda_pad ....................   Passed    0.60 sec
      Start 58: test_cuda_pooling
58/71 Test #58: test_cuda_pooling ................   Passed    3.35 sec
      Start 59: test_cuda_reduce
59/71 Test #59: test_cuda_reduce .................   Passed    3.54 sec
      Start 60: test_cuda_reshape
60/71 Test #60: test_cuda_reshape ................   Passed    3.52 sec
      Start 61: test_cuda_resize
61/71 Test #61: test_cuda_resize .................   Passed    4.69 sec
      Start 62: test_cuda_sendrecv
62/71 Test #62: test_cuda_sendrecv ...............   Passed   17.25 sec
      Start 63: test_cuda_slice
63/71 Test #63: test_cuda_slice ..................   Passed    3.09 sec
      Start 64: test_cuda_softmax
64/71 Test #64: test_cuda_softmax ................   Passed    3.49 sec
      Start 65: test_cuda_split
65/71 Test #65: test_cuda_split ..................   Passed    0.59 sec
      Start 66: test_cuda_transpose
66/71 Test #66: test_cuda_transpose ..............   Passed    3.39 sec
      Start 67: test_cuda_unary
67/71 Test #67: test_cuda_unary ..................   Passed    3.45 sec
      Start 68: test_cuda_unary_fp16
68/71 Test #68: test_cuda_unary_fp16 .............   Passed    0.01 sec
      Start 69: test_cuda_where
69/71 Test #69: test_cuda_where ..................   Passed    0.63 sec
      Start 70: test_perfengine
70/71 Test #70: test_perfengine ..................   Passed    3.51 sec
      Start 71: test_nccl_comm
71/71 Test #71: test_nccl_comm ...................   Passed    7.30 sec

100% tests passed, 0 tests failed out of 71`

英伟达平台测试通过。

**Comment by bitzyz** [comment]
image
mlu测试无误

**Comment by bitzyz** [comment]
昆仑平台测试无误

**Review by wanghailu0717 — APPROVED** [review]

**Review by bitzyz — CHANGES_REQUESTED** [review]

File: examples/distributed/parallel_opt.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def shard_transpose(node: NodeProto):
             place[node.output[0]] = Shard(list(perm).index(plc.dim))
 
     def shard_node(node: NodeProto):
-        if node.op_type in ["Relu", "Tanh", "Softmax"]:
+        if node.op_type in ["Relu", "Tanh", "Softmax", "Cast"]:
```
[the remark this answers is not in the document]
Comment (kilinchange) [reply examples/distributed/parallel_opt.py:140]: 这部分改动是适配 llama fp16 分布式的，还是合一下吧，int8 的部分我没加进来。

File: src/kernels/cuda/gather.cu [hunk 2]
```diff
@@ -28,27 +28,32 @@ __device__ T gatheredOffset2Offset(int gOffset,
...
 
-template <typename T>
-__global__ void _gather_kernel(float *in, float *out,
+template <typename dataT, typename T>
+__global__ void _gather_kernel(dataT *in, dataT *out,
                                infini::GatherMetaData metaData, size_t num) {
     T tid = threadIdx.x + blockIdx.x * blockDim.x;
-    int stride = blockDim.x * gridDim.x;
-    while (tid < num) {
+    if (tid < num) {
```
[the remark this answers is not in the document]
Comment (kilinchange) [reply src/kernels/cuda/gather.cu:35]: @OdinaryWord

File: src/kernels/cuda/all_reduce.cc [hunk 3]
```diff
@@ -13,15 +13,20 @@ class AllReduceNCCL : public CudaKernelWithoutConfig {
         auto context = dynamic_cast<const CudaRuntimeObj *>(_context);
         void *input = op->getInputs(0)->getRawDataPtr<void *>();
         void *output = op->getOutput()->getRawDataPtr<void *>();
-        IT_ASSERT(op->getDType() == DataType::Float32);
+        ncclDataType_t ncclType = ncclFloat;
+        if (op->getDType() == DataType::Float16) {
+            ncclType = ncclFloat16;
+        } else if (op->getDType() == DataType::Int8) {
+            ncclType = ncclInt8;
+        }
```
[the remark this answers is not in the document]
Comment (kilinchange) [reply src/kernels/cuda/all_reduce.cc:23]: fixed

File: include/core/operator.h [hunk 4]
```diff
@@ -90,6 +90,8 @@ class OperatorObj : public Object {
     OpType getOpType() const { return type; }
     // HACK: set correct data type
     DataType getDType() const { return getInputs(0)->getDType(); }
+    DataType getInDType() const { return getInputs(0)->getDType(); }
```
[the remark this answers is not in the document]
Comment (kilinchange) [reply include/core/operator.h:93]: fixed

File: src/kernels/cuda/gather.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (kilinchange) [reply src/kernels/cuda/gather.cu:35]: 之前的写法逻辑是不会进第二次while的，也就是说stride算着没意义。因为blockDim.x * gridDim.x >= num。

[the remark this answers is not in the document]
Comment (OdinaryWord) [reply src/kernels/cuda/gather.cu:35]: 之前的写法逻辑是不会进第二次while的，也就是说stride算着没意义。因为blockDim.x * gridDim.x >= num。

**Review by bitzyz — APPROVED** [review]

**Review by whjthu — APPROVED** [review]

**Pull request merged**